src/audio/selector/selector.c

```diff
@@ -152,6 +94,111 @@ static void selector_free(struct comp_dev *dev)
+	 * cd->config.in/out_channels_count = 0 means that it can vary.
+	 */
+	if (dev->direction == SOF_IPC_STREAM_PLAYBACK) {
+		/* fetch sink buffer for playback */
+		buffer = list_first_item(&dev->bsink_list, struct comp_buffer,
+					 source_list);
+		if (cd->config.in_channels_count &&
+		    cd->config.in_channels_count != params->channels) {
+			trace_selector_error_with_ids(dev,
+						      "selector_verify_params() error: src in_channels_count does not match pcm channels"
+						     );
+			return -EINVAL;
+		}
+		in_channels = cd->config.in_channels_count;
+
+		/* if cd->config.out_channels_count are equal to 0
+		 * (it can vary), we set params->channels to sink buffer
+		 * channels, which were previosly set in
+		 * pipeline_comp_hw_params()
+		 */
+		out_channels = cd->config.out_channels_count ?
+			cd->config.out_channels_count : buffer->channels;
+		params->channels = out_channels;
+	} else {
+		/* fetch source buffer for capture */
+		buffer = list_first_item(&dev->bsource_list, struct comp_buffer,
+					 sink_list);
+		if (cd->config.out_channels_count &&
+		    cd->config.out_channels_count != params->channels) {
+			trace_selector_error_with_ids(dev,
+						      "selector_verify_params() error: src in_channels_count does not match pcm channels"
+						     );
+			return -EINVAL;
+		}
+		out_channels = cd->config.out_channels_count;
+
+		/* if cd->config.in_channels_count are equal to 0
+		 * (it can vary), we set params->channels to source buffer
+		 * channels, which were previosly set in
+		 * pipeline_comp_hw_params()
+		 */
+		in_channels = cd->config.in_channels_count ?
+			cd->config.in_channels_count : buffer->channels;
+		params->channels = in_channels;
+	}
+
+	/* Set buffer params */
+	buffer_set_params(buffer, params, BUFFER_UPDATE_FORCE);
+
+	/* set component period frames */
+	component_set_period_frames(dev, sinkb->rate);
+
+	/* verify input channels */
+	if (in_channels != SEL_SOURCE_2CH && in_channels != SEL_SOURCE_4CH) {
+		trace_selector_error_with_ids(dev,
+					      "selector_verify_params() error: in_channels = %u"
+					      , in_channels);
+		return -EINVAL;
+	}
+
+	/* verify output channels */
+	if (out_channels != SEL_SINK_1CH && out_channels != SEL_SINK_2CH &&
+	    out_channels != SEL_SINK_4CH) {
+		trace_selector_error_with_ids(dev,
+					      "selector_verify_params() error: out_channels = %u"
+					      , out_channels);
+		return -EINVAL;
+	}
+
+	/* verify proper channels for passthrough mode */
+	if (out_channels != SEL_SINK_1CH && in_channels != out_channels) {
+		trace_selector_error_with_ids(dev,
+					      "selector_verify_params() error: in_channels = %u, out_channels = %u"
+					      , in_channels, out_channels);
+		return -EINVAL;
+	}
+
+	if (cd->config.sel_channel > (SEL_SOURCE_4CH - 1)) {
+		trace_selector_error_with_ids(dev,
+					      "selector_verify_params() error: ch_idx = %u"
+					      , cd->config.sel_channel);
+		return -EINVAL;
+	}
+
```

Comment:
above comment on ```switch``` applies down to here.

Reply:
@lgirdwood 
Do you also want switch statement for ```config.sel_channel```? It seems to me that it would be less readable, since channel is indexed from 0.

Reply:
no, this looks fine.

Reply:
no